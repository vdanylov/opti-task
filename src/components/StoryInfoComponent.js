import React, { Component } from 'react'
import { Button, Dialog, DialogContent, DialogActions } from 'react-mdl'

export default class StoryInfoComponent extends Component {

  state = {
    openDialog: false
  }

  handleOpenDialog = () => {
    this.setState({
      openDialog: true
    });
  }

  handleCloseDialog = () => {
    this.setState({
      openDialog: false
    });
  }

  renderIframe = url => {
    return {__html: `<iframe src=${url} width="100%" height="550px"></iframe>`};
  }

  render() {
    
    const { 
      story: { url, by, title, type}, 
      closeShowInfo 
    } = this.props;

    const { openDialog } = this.state;

    return (
      <div>
        <Dialog open={openDialog} style={{width: "80%", height: "80%"}}>
          <DialogContent>
            <div dangerouslySetInnerHTML={this.renderIframe(url)} />
          </DialogContent>
          <DialogActions>
            <Button type='button' onClick={this.handleCloseDialog}>Close</Button>
          </DialogActions>
        </Dialog>
        <div>
          <h2><b>Additional story info</b></h2>
          <p><b>Autor:</b> {by}</p>
          <p><b>Title:</b> {title}</p>
          <p><b>Title:</b> {type}</p>
          <Button colored onClick={this.handleOpenDialog}>Read</Button>
          <Button colored onClick={closeShowInfo}>Close</Button>        
        </div>
      </div>
    )
  }
}
