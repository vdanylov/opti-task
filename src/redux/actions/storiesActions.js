import axios from 'axios'

export const getStories = () => {
    return axios(`https://hacker-news.firebaseio.com/v0/topstories.json?print=pretty`, {
        method: "GET",
        credentials: "include",
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json"
        }
    })
    .then(res => {
        if (res.status === 200) {
            return res.data;
        } else {
            Promise.reject(res);
        }        
    })
    .then(res => {
        const stories = res.map(story => getSingleStory(story));
        return axios.all(stories);
    })
    .then(res => {
        return res.map(story => story.data)
    })
    .catch(error => {
        // console.log(error);
    })
};


const getSingleStory = id => {
    return axios(`https://hacker-news.firebaseio.com/v0/item/${id}.json?print=pretty`, {
        method: "GET",
        credentials: "include",
        headers: {
            "Accept": "application/json",
            "Content-Type": "application/json"
        }
    })
}