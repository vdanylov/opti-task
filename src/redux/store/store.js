import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import mainReducer from '../reducers/mainReducer'
import storiesSaga from '../sagas/mainSaga'

const sagaMiddleware = createSagaMiddleware();

const store = createStore(
    mainReducer, {},
    compose(
        applyMiddleware(sagaMiddleware),
        window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
    )
)

sagaMiddleware.run(storiesSaga);

export default store;
