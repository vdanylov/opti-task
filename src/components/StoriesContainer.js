import React, { Component } from 'react'
import constants from '../redux/constants/index'
import { connect } from 'react-redux'
import StoriesDataTable from './StoriesDataTable'
import StoryInfoComponent from './StoryInfoComponent'

class StoriesContainer extends Component {

    state = {
        showInfo: false,
        storyId: null
    }

    componentDidMount = () => this.props.dispatch({ type: constants.GET_STORIES})
    
    showMoreInfo = (storyId,e) => {
        e.preventDefault();
        this.setState((state, props) => ({ 
            showInfo: true,
            storyId
        }));
    }

    closeShowInfo = () => this.setState({ showInfo: false});

    render() {
        const { stories: { isLoading, stories, error } } = this.props;
        const { storyId, showInfo } = this.state;
        const story = stories.find(story => story.id === storyId);

        return (
        <div>
            {!isLoading ?
                <div>
                    {!showInfo &&
                        <StoriesDataTable 
                            stories={stories}
                            showMoreInfo={this.showMoreInfo}
                        />                    
                    }
                    {(showInfo && story) && 
                        <StoryInfoComponent 
                            story={story} 
                            closeShowInfo={this.closeShowInfo}                        
                        />}
                </div> : <div>Loading...</div>
            }
        </div>
        )
    }
}

const mapStateToProps = state => ({
    stories: state.storiesReducer
})

const mapDispatchToProps = dispatch => ({
    dispatch
})

export default connect(mapStateToProps, mapDispatchToProps)(StoriesContainer);