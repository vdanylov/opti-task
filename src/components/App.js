import React, { Component } from 'react';
import logo from '../styles/logo.svg';
import '../styles/App.css';
import { Provider } from 'react-redux'
import store from '../redux/store/store'
import 'react-mdl/extra/material.css';
import 'react-mdl/extra/material.js';
import StoriesContainer from './StoriesContainer'
import { Grid, Cell } from 'react-mdl'


class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <div style={{width: '80%', margin: 'auto'}}>
          <Grid>
            <Cell col={12}>
              <ReactHeader />
            </Cell>
          </Grid>
          <Grid>
          <div style={{margin: '0 auto'}}>
              <StoriesContainer />
          </div>
        </Grid>
        </div>
      </Provider>    
    );
  }
}

const ReactHeader = () => {
  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <h1 className="App-title">Top stories on HackerNews</h1>
      </header>        
    </div>
  );
};

export default App;
