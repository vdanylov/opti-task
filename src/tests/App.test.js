import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import App from '../components/App';
import { getStories } from '../redux/actions/storiesActions'
import StoriesContainer from '../components/StoriesContainer'
import StoryInfoComponent from '../components/StoryInfoComponent'
import store from '../redux/store/store'

const story = {
  by: 'jaytaylor',
  descendants: 44,
  id: 15955711,
  kids: [
    15955952,
    15955803,
    15955887,
    15956333,
    15956160,
    15955939,
    15956471,
    15956206,
    15956324,
    15956350,
    15955771
  ],
  score: 144,
  time: 1513632819,
  title: 'Older Adults\' Forgetfulness Tied to Faulty Brain Rhythms in Sleep',
  type: 'story',
  url: 'http://npr.org/sections/health-shots/2017/12/18/571120472/older-adults-forgetfulness-tied-to-faulty-brain-rhythms-in-sleep'
}

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});

describe('<App />', () => {
  expect.addSnapshotSerializer({
    test: val => val.by && val.score && val.title,
    print: val => `${val.by}  ${val.score}  ${val.title}`
  })

  it('Request data match test', () => {
    expect(getStories()).toMatchSnapshot();
  })

  it('Renders StoriesContainer', () => {
    const component = renderer.create(
      <StoriesContainer store={store} />
    )
    expect(component).toMatchSnapshot();
  })

  it('Renders StoriesContainer', () => {
    const component = renderer.create(
      <StoriesContainer store={store} />
    )
    expect(component).toMatchSnapshot();
  })

  it('Renders StoryInfoComponent', () => {
    const component = renderer.create(
      <StoryInfoComponent story={story} closeShowInfo={() => {}} />
    )
    expect(component).toMatchSnapshot();
  })
})

