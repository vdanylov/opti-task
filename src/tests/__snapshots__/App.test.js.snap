// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Renders StoriesContainer 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`<App /> Renders StoriesContainer 2`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`<App /> Request data match test 1`] = `Promise {}`;
